src/tools/roc_recv/main.cpp

```diff
@@ -116,6 +122,12 @@ int main(int argc, char** argv) {
         }
     }
 
+    if (!args.max_frame_size_given) {
+        context_config.max_frame_size = sizeof(audio::sample_t)
+            * receiver_config.common.output_sample_spec.ns_2_samples_overall(
+                io_config.frame_length);
+    }
```

Comment:
There are 2 problems here:

1. It's too late to update max_frame_size, we already created context from older version of context_config.

2. It's too early to invoke ns_2_samples_overall(), because later we can call

    ```
    receiver_config.common.output_sample_spec.set_sample_rate(
        output_sink->sample_spec().sample_rate());
    ```

    which will set pipeline rate to device rate, which will change result of ns_2_samples_overall().

I think the desired sequence would be:

* fill configs
* open devices
* update configs
* open context and receiver

Reply:
Can't figure this one out. There seems to be some cyclical dependency.

* The output_sink requires a context to already exist (for the arena).

Reply:
Oh, I see. Seems the problem is even deeper: opening device needs frame pool, frame pool needs frame size, frame size needs rate, which needs to open device.

I don't see any simple solution for this. We could do something tricky like deferring frame pool initialization, but I think complicating code for this small feature does worth it.

I suggest a simpler solution then:

* calculate max_frame_size manually, without using information from receiver config and device 
* if --rate is specified, use its value for calculations
* if --rate is not specified, use 48000 (in 99% cases rate will be below this value)
* for channel count, use 2, as currently we don't allow to configure device channel count
* we'll add an option for channel count in next release (0.4 I think), then we'll use specified channel count instead of 2 if that option is given

Reply:
OK. I've had to move the ContextConfig and the Context creation further down. I've then taken the rate and frame_length from io_config, if they were provided - otherwise defaults. And then used that to calculate the max_frame_size.

Reply:
Nice!